branches/0.9/ddsnap/man/ddsnap-sb.8
.TH ddsnap-sb 8 "Apr 29, 2008" "Linux"
.SH NAME
ddsnap-sb \- ddsnap helper program for upgrading ddsnap ondisk superblock structure.

\fB
.SH SYNOPSIS
.B ddsnap-sb
.br
.SH DESCRIPTION
\fBddsnap-sb\fP is used to upgrade the ondisk superblock of an existing ddsnap volume 
to the latest version.

.SH BUGS
Please report bugs at \fBhttp://code.google.com/p/zumastor\fP or mail them to \fBzumastor@googlegroups.com\fP.

.SH VERSION
This man page is current for version 0.8 of \fBddsnap-sb\fP.

.SH AUTHORS
.TP
zumastor-team

.SH THANKS
.TP
Thanks to Google, Red Hat and Sistina Software for supporting this work.  Special thanks to: Mike Todd, Joseph Dries and Matthew O'Keefe.
.TP
The home page of \fBddsnap-sb\fP is \fBhttp://zumastor.org\fP.  This site may cover questions unanswered by this manual page.  Mailing lists for support and development are available at zumastor@googlegroups.com


trunk/Changelog
Release 0.4-pre7:

There is much stability and cleanup work in this release, including major
reworking of the automatic snapshot deletion model to support automatic
deletion of snapshots even if they are still mounted (geewhiz).  Also lots
of work aimed at being able to operate reasonably well with a snapshot
store smaller than the origin volume.  The dominating item for this cycle
was fixing the congestion_wait deathtrap on recursive memory writeout, a
longstanding bug we inherited from core kernel.

New features:

  * Progress interface added to ddsnap transmit, suitable for use by shell
    scripts, or interpreted or compiled programs.

  * Extended functionality of zumastor snapshot to allow the admin to create
    an arbitrary snapshot that is not part of the regular snapshot schedule
    and will not be automatically deleted

  * ddsnap magic number format improved to incorporate the date of most
    recent incompatible change

  * A nascent automated test harness, not yet functional has appeared

  * Incremental used chunk statistics are now recorded in the journal
    commit block so that they are always correct on journal replay, which
    avoids a full scan of the btree on unclean restart

  * A snapshots that is in use or even mounted may now be automatically
    deleted if the snapshot store is full and somebody is trying to write
    to the origin.  Much better than failing the origin write.  This is
    termed "quashing" a snapshot, and the snapshot so abused will cause an
    IO error if read or written to by the unfortunate filesystem mount,
    but will not cause any system code to fault, oops, hang or otherwise
    behave.  Probably.

  * We now have an automated script to build debian packages

  * Concept of "full volume delta" introduced so that an entire volume may
    be sent downstream using the streaming delta format, which includes
    compression of the volume data

  * Initial replication is now possible even when downstream snapshot store
    is (much) smaller than the origin volume.  This is accomplished using
    a full volume replication cycle without any snapshot set downstream,
    after which normal, snapshot-based incremental replication begins

  * Quota support mount option is now enabled by default.  (Later this will
    turn into arbitrary per-volume mount option support)

  * Default chunk size changed from 64K to 16K to improve because 64K wastes
    too much snapshot store space due to preserving unchanged origin data

  * No longer any need to keep the primordial snapshot 0 around forever,
    which saves the snapshot store from eventually inhaling the entire
    origin volume under less than extreme churn patterns

  * If replication fails we now retry in case it was just a transient error.
    Next prerelease we will add functionality to distinguish between
    transient and serious replication errors, to decide whether to initiate
    a full volume replication cycle or just to retry the replication as now.

  * we have put the snapshot priorities to use for the first
    time.  If a snapshot priority is set to "127" it will be unquashable,
    causing origin writes to fail rather than have it quashed.  This is
    used to weigh the preference of origin write failures rather than
    snapshot quashing downstream where we prefer to keep exporting rather
    than getting up to date.  This should become optional for 0.5.

  * Added options to a number of commands

  * "noinit" made default for "zumastor define volume".  optional parameter
    parsing for this command has been switched to getopt.  "nosnap" option
    has been removed due to support full volume replication (now the only
    way)

  * Added a proc interface under /proc/fs/ddsnap so we can see immediately
    how many IOs are in flight for each client.

Bug fixes:

  * Added throttling of IO submitters in order to place a bound on IOs
    in flight, and thus place a bound on memory usage to service them,
    which is required in order to do the proper fix to memory writeout
    recursion to replace the broken kernel congestion_wait mechanism.
    Side effect is improved latency of block device read operations.

  * lockups to the nasty kernel vm congestion_wait mechanism now have
    workarounds involving the semi-functional PF_LESS_THROTTLE mechanism.
    Much more to do here.

  * ddsnap virtual devices used to allow hundreds of thousands of write
    operations to be in flight to the snapshot server at the same time,
    causing severe latency problems under some conditions, and leaving
    the door open to low memory deadlock.  Our virtual devices now
    throttle themselves to allow a maximum of a thousand non-vm-driven
    writes to be in flight at once

  * Much useless server and client log output that was creating huge log
    files is removed, keeping the log output that proved useful.

  * Rationalized the algorithms for checking consistency of upstream and
    downstream database at replication initiation time

  * zumastor stop in the middle of replication failed to remove the
    associated snapshot devmapper device, fixed.

  * "zumastor stop source" now stops the ddsnap listen daemon as it should

  * ddsnap now checks for a superblock magic before doing any operation on
    the snapshot store.  Before this, ddsnap would happily try to allocate
    a silly amount of buffer memory if trying to initialize on an obsolete
    snapshot store.

  * Much cleanup and refactoring of functions, files and interfaces so we
    can leave that brown paper bag in the cupboard when new visitors come
    to look at the code

  * Zumastor status now does a scan of the freespace bitmaps to see if they
    match the incremental free chunk statistics

  * Bugs in free chunk counting fixed so that used and free counts match
    each other, and match the actual free space from scanning the bitmaps

  * Fixed many bugs related to automatic deletion of snapshots on snapshot
    store full


ddsnap/ddsnap-sb.c
/*
 * This program is used to upgrade the ondisk ddsnap superblock.
 * You should add an upgrade function here and change the SB_MAGIC in ddsnapd.c
 * everytime you change the ddsnap ondisk superblock structure.
 */
#include <stdlib.h>
#include <string.h>
#include <stddef.h>
#include <stdio.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/types.h>
#include <time.h>
#include "diskio.h"
#include "buffer.h"
#include "trace.h"
#include "ddsnap.h"
#include "dm-ddsnap.h"

// FIXME: the following micros are copied from ddsnapd.c, move them to a header file
#define MAX_SNAPSHOTS 64
#define SB_SECTOR 8		/* Sector where superblock lives.     */
#define SB_SECTORS 8		/* Size of ddsnap super block in sectors */

#define SB_MAGIC_070604 { 't', 'e', 's', 't', 0xdd, 0x07, 0x06, 0x04 }	/* 0.4 - 0.7 release sb magic */
#define SB_MAGIC_080325 { 't', 'e', 's', 't', 0xdd, 0x08, 0x03, 0x25 }	/* 0.8 release sb magic */
#define SB_MAGIC { 't', 'e', 's', 't',  0xdd, 0x07, 0x06, 0x04 }	/* date of latest incompatible sb format */

struct allocspace_img {
	sector_t bitmap_base;	/* Allocation bitmap starting sector. */
	sector_t chunks;	/* if zero then snapdata is combined in metadata space */
	sector_t freechunks;
	chunk_t last_alloc;	/* Last chunk allocated.              */
	u64 bitmap_blocks;	/* Num blocks in allocation bitmap.   */
	u32 allocsize_bits;	/* Bits of number of bytes in chunk. */
};

struct disksuper_04 {		/* the subfix is the data of latest incompatible sb format: yymmdd */
	typeof((char[])SB_MAGIC) magic;
	u64 create_time;
	sector_t etree_root;	/* The b-tree root node sector.       */
	sector_t orgoffset, orgsectors;
	u64 flags;
	u64 deleting;
	struct snapshot_04 {
		u32 ctime;	// upper 32 bits are in super create_time
		u32 tag;	// external name (id) of snapshot
		u16 usecount;	// persistent use count on snapshot device
		u8 bit;		// internal snapshot number, not derived from tag
		s8 prio;	// 0=normal, 127=highest, -128=lowest
		char reserved[4];	/* adds up to 16 */
	} snaplist[MAX_SNAPSHOTS];	// entries are contiguous, in creation order
	u32 snapshots;
	u32 etree_levels;
	s32 journal_base;	/* Sector offset of start of journal. */
	s32 journal_next;	/* Next chunk of journal to write.    */
	s32 journal_size;	/* Size of the journal in chunks.     */
	u32 sequence;		/* Latest commit block sequence num.  */
	struct allocspace_img metadata, snapdata;
};

struct snapshot_06 {
	u32 ctime;		// upper 32 bits are in super create_time
	u32 tag;		// external name (id) of snapshot
	u16 usecount;		// persistent use count on snapshot device
	u8 bit;			// internal snapshot number, not derived from tag
	s8 prio;		// 0=normal, 127=highest, -128=lowest
	u64 sectors;		// sectors of the snapshot
};

struct disksuper_06 {
	typeof((char[])SB_MAGIC) magic;
	u64 create_time;
	sector_t etree_root;	/* The b-tree root node sector.       */
	sector_t orgoffset, orgsectors;
	u64 flags;
	u64 deleting;
	struct snapshot_06 snaplist[MAX_SNAPSHOTS];	// entries are contiguous, in creation order
	u32 snapshots;
	u32 etree_levels;
	s32 journal_base;	/* Sector offset of start of journal. */
	s32 journal_next;	/* Next chunk of journal to write.    */
	s32 journal_size;	/* Size of the journal in chunks.     */
	u32 sequence;		/* Latest commit block sequence num.  */
	struct allocspace_img metadata, snapdata;
};

static int sb_print(int metadev)
{
	struct disksuper_06 sb;	/* latest disksuper structure */
	int err;
	if ((err = diskread(metadev, &sb, sizeof(sb), SB_SECTORS << SECTOR_BITS)) < 0) {
		warn("Unable to write superblock: %s", strerror(errno));
		return err;
	}
	printf("creation_time %llu, flags %llu\n", (unsigned long long)sb.create_time, (unsigned long long)sb.flags);
	for (int i = 0; i < sb.snapshots; i++) {
		printf("ctime %u, tag %u, usecount %u, bit %u, prio %d",
		       sb.snaplist[i].ctime, sb.snaplist[i].tag,
		       sb.snaplist[i].usecount, sb.snaplist[i].bit, sb.snaplist[i].prio);
		printf(", sectors %llu", sb.snaplist[i].sectors);
		printf("\n");
	}
	printf("snapshots %d\n", sb.snapshots);
	return 0;
}

/* upgrade from 0.4 release to 0.6 release */
static int upgrade_04(int metadev, char *magic)
{
	struct disksuper_04 oldsb;
	struct disksuper_06 sb;
	int err;

	warn("upgrade sb from 0.4 format to 0.6 format");
	if ((err = diskread(metadev, &oldsb, sizeof(struct disksuper_04), SB_SECTORS << SECTOR_BITS)) < 0) {
		warn("Unable to read old superblock: %s", strerror(errno));
		return err;
	}

	memcpy(&sb, &oldsb, ((char *)&oldsb.snaplist - (char *)&oldsb));
	for (int i = 0; i < oldsb.snapshots; i++) {
		memcpy(&sb.snaplist[i], &oldsb.snaplist[i], sizeof(struct snapshot_06));
		sb.snaplist[i].sectors = oldsb.orgsectors;
	}
	memcpy(&sb.snapshots, &oldsb.snapshots,
	       ((char *)&oldsb + sizeof(struct disksuper_04) - (char *)&oldsb.snapshots));

	if ((err = diskwrite(metadev, &sb, sizeof(struct disksuper_06), SB_SECTORS << SECTOR_BITS)) < 0) {
		warn("Unable to write superblock: %s", strerror(errno));
		return err;
	}
	return err;
}

/* upgrade from 0.6 release to 0.8 */
static int upgrade_06(int metadev, char *magic)
{
	int err;
	warn("just overwrite the sb magic");
	memcpy(magic, (char[])SB_MAGIC_080325, sizeof((char[])SB_MAGIC_080325));
	if ((err = diskwrite(metadev, magic, sizeof((char[])SB_MAGIC_080325), SB_SECTORS << SECTOR_BITS)) < 0)
		warn("Unable to write superblock: %s", strerror(errno));
	return err;
}

int main(int argc, char *argv[])
{
	typeof((char[])SB_MAGIC) magic;
	int metadev, err, release4 = 0;

	if (argc < 2)
		error("usage: %s upgrade <metadev> [0.4]", argv[0]);
	if ((metadev = open(argv[1], O_RDWR | O_SYNC)) < 0)
		error("unable to open the metadata device, err %s\n", strerror(errno));
	if ((err = diskread(metadev, magic, sizeof(magic), SB_SECTORS << SECTOR_BITS)) < 0) {
		warn("Unable to read superblock magic: %s", strerror(errno));
		goto out;
	}
	warn("sb magic %c%c%c%c %2x-%2x-%2x", magic[0], magic[1], magic[2],
	     magic[3], (unsigned)magic[5], (unsigned)magic[6], (unsigned)magic[7]);

	/* forgot to change sb magic number in 0.6 release, so need special handling between 0.4 to 0.6 */
	if (argc > 2 && !strcmp(argv[2], "0.4") && !memcmp(magic, (char[])SB_MAGIC_070604, sizeof(magic)))
		release4 = 1;

	while (!err && (memcmp(magic, (char[])SB_MAGIC, sizeof(magic)) || release4)) {
		if (!memcmp(magic, (char[])SB_MAGIC_070604, sizeof(magic))) {
			if (release4) {
				upgrade_04(metadev, magic);
				release4 = 0;
			} else
				upgrade_06(metadev, magic);
		} else if (!memcmp(magic, (char[])SB_MAGIC_080325, sizeof(magic))) {
			printf("magic number is up to date, do nothing\n");
			//upgrade_08(metadev, magic);
		} else {
			printf("superblock magic %s not supported\n", magic);
			err = -EINVAL;
		}
	}
	sb_print(metadev);
out:
	close(metadev);
	return err;
}
